server/src/server.js

```diff
@@ -0,0 +1,156 @@
+const { v4: uuidv4 } = require("uuid");
+const express = require("express");
+const cookieSession = require("cookie-session");
+const datastore = require("./datastore");
+const app = express();
+
+const KIND_EVENT = "Event";
+const URL_PARAM_EVENT_ID = `eventID`;
+
+const ERROR_BAD_DB_INTERACTION = "BAD_DATABASE";
+const ERROR_INVALID_EVENT_ID = "INVALID_EVENT_ID";
+const ERROR_BAD_UUID = "BAD_UUID";
+
+// Parse request bodies with the json content header into JSON
+app.use(express.json());
+
+app.use(
+  cookieSession({
+    name: "session",
+    keys: ["secret1"],
+    maxAge: 365 * 24 * 60 * 60 * 1000, // 1 year
+  })
+);
+
+app.use((req, res, next) => {
+  if (req.session.isNew) {
+    // Don't worry about collision
+    const uuid = uuidv4();
+    req.session.userID = uuid;
+  } else {
+    if (
+      !/^[a-z0-9]{8}-[a-z0-9]{4}-4[a-z0-9]{3}-[a-z0-9]{4}-[a-z0-9]{12}$/.test(
+        req.session.userID
+      )
+    ) {
+      return res.status(500).json({
+        status: 500,
+        error: { type: ERROR_BAD_UUID },
+      });
+    }
+  }
+  next();
+});
+
+/**
+ *  Middleware that can be used on routes that match `/:${URL_PARAM_EVENT_ID}/...`
+ *  This will fetch the event from Datastore and attach it and its key to
+ *  req.event and req.datastoreKey respectively.
+ *
+ *  It will return an error if, for any reason, the event could not be found.
+ */
+async function getEvent(req, res, next) {
+  const key = datastore.key([
+    KIND_EVENT,
+    // If the url is: /1/details
+    // then `req.params[URL_PARAM_EVENT_ID]` will be 1
+    parseInt(req.params[URL_PARAM_EVENT_ID]),
+  ]);
+  datastore
+    .runQuery(datastore.createQuery(KIND_EVENT).filter("__key__", "=", key))
+    .then((results) => {
+      // results[0] is the actual returned entities
+      // results[1] is metadata about the request (e.g are there additional results)
+      const events = results[0];
+      if (events.length === 0) {
+        res.status(400).json({
+          status: 400,
+          error: {
+            type: ERROR_INVALID_EVENT_ID,
+          },
+        });
+      } else {
+        const event = events[0];
+        if (
+          event === undefined ||
+          event === null ||
+          typeof event !== "object"
```

Comment:
oh boy, gotta love js :D

Reply:
Haha, this should never happen, but I added in the check anyway just to be overly defensive :p